gnome-initial-setup/gis-page-util.c

```diff
@@ -446,3 +446,68 @@ gis_page_util_show_factory_dialog (GisPage *page)
 
   g_free (version);
 }
+
+typedef struct _DemoButtonData {
+  GtkDialog *demo_dialog;
+  GisDriver *driver;
+} DemoButtonData;
+
+static DemoButtonData *
+demo_button_data_new (GtkDialog *dialog, GisDriver *driver)
+{
+  DemoButtonData *data = (DemoButtonData *) g_slice_new (DemoButtonData);
+  data->demo_dialog = dialog;
+  data->driver = g_object_ref (driver);
+
+  return data;
+}
+
+static void
+demo_button_data_free (gpointer data, GClosure *closure)
+{
+  DemoButtonData *db_data;
+
+  g_clear_object (&db_data->driver);
+  g_slice_free (DemoButtonData, db_data);
+}
+
+static void
+pressed_demo_button (GtkWidget *widget, gpointer user_data)
+{
+  DemoButtonData *data = (DemoButtonData *) user_data;
+
+  gis_driver_enter_demo_mode (data->driver);
+  gtk_window_close (GTK_WINDOW (data->demo_dialog));
```

Comment:
You could grab the toplevel from the button and avoid the aux structure, if you simply pass the driver object around instead